5단계 - 로또(수동)

victor-jo opened the pull request:
5단계 - 로또(수동) 리뷰 요청드립니다~!
잘 부탁드리겠습니다

victor-jo committed:
style(lottoScanner): 로또 스캐너 메소드 재활용을 위해 이름 변경

before:
    scanWinningNumbers 에서
after:
    scanManyNumbers 으로

victor-jo committed:
docs(README_LOTTO): 로또 구매자 기능 추가

- 로또 구매자가 직접 로또 STATIC 앱을 통해 당첨 확인
- 로또 구매자가 스스로 수동을 구매하는 기능
- 로또 구매자가 스스로 자동을 구매하는 기능
- 자신이 가지고 있는 로또들 중 당첨이 있는지 확인하는 기능

victor-jo committed:
refactor(lottoStore): 로또 상점에 구매 전략을 말하는 메시지 추가

- 로또 상점에 자동으로 산다고 말하면 자동으로 삼
- 로또 상점에 수동으로 산다고 말하면 수동으로 삼

victor-jo committed:
refactor(lottoStore): 로또 상점에 구매 개수를 말하는 메시지 추가

- 구매 개수를 초과할 경우 최대를 산다
- 구매 개수 안쪽으로는 살 수 있도록 한다

victor-jo committed:
feat(lottoBuyer): 로또 구매자 객체 추가

- 로또 구매자는 수동구매, 자동구매를 선택하여 산다
- 내가 산 모든 로또를 모아놓았다가 제출한다
- 구매한 로또 중 수동이 몇 개인지 알려준다
- 구매한 로또 중 자동이 몇 개인지 알려준다

victor-jo committed:
refactor(lottoApp): 로또 구매자 객체 적용

- 로또 수동구매 기능 추가

joojimin commented on src/main/java/lotto/LottoApp.java:
```diff
@@ -12,35 +11,49 @@
-        for (Lotto lotto : lottoList) {
-            print(lotto.toString());
+        print(BUY_MANUAL_NUMBER);
+        for (int i = 0; i < buyManualCount; i++) {
```
해당 로직도 LottoBuyer에 넣을 수 있지않을까요~? 
더 깔끔하게 처리가 될 것 같습니다!

joojimin commented on src/main/java/lotto/LottoBuyer.java:
```diff
@@ -0,0 +1,51 @@
+
+public class LottoBuyer {
+    private final PayAmount payAmount;
+    private final List<Lotto> lottoList = new ArrayList<>();
```
일급컬렉션은 대상 컬렉션외 다른 멤버변수를 갖지 않는걸 의미해요! lottoList를 일급컬렉션으로 만들고, 안으로 가져갈 수 있는 로직은 분리하여 LottoBuyer와 분리해보면 어떨까요?

joojimin commented on src/main/java/lotto/LottoStore.java:
```diff
@@ -2,21 +2,31 @@
 public class LottoStore {
 
     private static final int LOTTO_AMOUNT = 1000;
+    private Supplier<LottoNumberGenerateStrategy> strategySupplier = RandomLottoNumberGenerateStrategy::new;
```
람다 활용 👍 👍 👍  전략패턴 활용 정말 좋은 것 같습니다~!!!!

joojimin commented on src/main/java/lotto/PayCount.java:
```diff
@@ -0,0 +1,22 @@
+public class PayCount {
+
+    private static final int INFINITE = -1;
+    private int count;
```
원시값 래핑 💯

joojimin reviewed (changes requested):
미션 진행 고생하셨습니다! 원준님~!!
몇가지 확인하고 넘어가면 좋을 법만 코멘트 드렸어요!
확인후 리뷰재요청 부탁드려요~!

victor-jo committed:
refactor(buyManualCount): 수동구매수 객체 추가

- 로또 수동구매 카운트 수 원시값 포장
- 수동번호 입력 행동과 수동구매 카운트 값을 받음
- 카운트 존재여부 판별 메시지 추가
- 수동번호 입력 행동에 따른 번호 생성 메시지 추가

victor-jo committed:
refactor(buyManualCount): 수동구매수 객체 리팩터링

- 수동번호 입력 행동 후 결과값 Arrays.asList 로 포장하여 전달하도록 메시지 수정

victor-jo committed:
refactor(lottoApp): 로또 앱 수동구매수 카운트 적용

- 수동구매수 카운트 적용

before:
    수동구매 원시값을 통해 앱에서 for 루프를 돌림
after:
    수동구매 원시값을 포장하여 메시지 처리함

victor-jo committed:
feat(lottoList): 로또 리스트 객체 추가

- 로또 리스트에서 번호 매치되는 값 받는 메시지 추가
- 총 수익 메시지 추가
- 보너스 번호와 당첨 로또를 기준으로 매치값 저장 메시지 추가

victor-jo committed:
refactor(lottoApp): 로또앱 리팩터링

- 로또 리스트 객체 적용
- 테스트에서만 사용하는 메시지 제거
- 당첨자 발표 통계 앱에서 로또 구매자를 받도록 수정

victor-jo committed:
refactor(lottoMatchTypeCounter): 로또매치 카운터 객체 추가

- 카운트 맵을 가지는 일급 컬렉션
- 로또매치타입에 대한 카운트값을 추가하거나 조회한다

victor-jo committed:
refactor(lottoList): 로또 리스트 리팩터링

- 로또 리스트 일급 컬렉션 (다른 변수를 가지지 않는) 만들기 위해 카운트맵 제외
- 로또 리스트 match 메시지에서 List 로 된 매치 타입 리스트 전달

victor-jo commented:
말씀하신 부분들 아래와 같이 처리하였습니다 확인 부탁드리겠습니다!

private final List<Lotto> lottoList = new ArrayList<>(); 을 LottoList 일급 컬렉션으로 분리시켰습니다!

분리 후 Count 값을 구하는 맵이 또 위치가 모호해서 LottoMatchTypeCounter 로 분리시켰습니다!

기존 로또 앱에서 수동 구매하는 for loop 부분은 BuyManualCount 를 원시값 포장하여 메시지 처리하였습니다

joojimin reviewed (approved):
피드백 반영 고생하셨습니다!!
훨씬 코드가 깔끔해진 것 같아요!!
다음 미션도 화이팅입니다!!
